datafusion/physical-expr/src/aggregate/count_distinct.rs

```diff
@@ -438,11 +446,292 @@ where
+                        return false;
+                    }
+                    inline == header.offset_or_inline
+                });
+
+                // Insert an entry for this value if it is not present
+                if entry.is_none() {
+                    let new_header = SSOStringHeader {
+                        hash,
+                        len: value.len(),
+                        offset_or_inline: inline,
+                    };
+                    self.map.insert_accounted(
+                        new_header,
+                        |header| header.hash,
+                        &mut self.map_size,
+                    );
+                }
+            }
+            // handle large strings
+            else {
+                // Check if the value is already present in the set
+                let entry = self.map.get_mut(hash, |header| {
+                    // if hash matches, must also compare the values
+                    if header.len != value.len() {
+                        return false;
+                    }
+                    // SAFETY: buffer is only appended to, and we correctly inserted values
+                    let existing_value =
+                        unsafe { self.buffer.as_slice().get_unchecked(header.range()) };
+                    value == existing_value
+                });
+
+                // Insert the value if it is not present
+                if entry.is_none() {
+                    // long strings are stored as a length/offset into the buffer
+                    let offset = self.buffer.len();
+                    self.buffer.append_slice(value);
+                    let new_header = SSOStringHeader {
+                        hash,
+                        len: value.len(),
+                        offset_or_inline: offset,
+                    };
+                    self.map.insert_accounted(
+                        new_header,
+                        |header| header.hash,
+                        &mut self.map_size,
+                    );
+                }
+            }
+        }
+    }
+
+    /// Converts this set into a StringArray of the distinct string values
+    fn into_state(self) -> StringArray {
+        // The map contains entries that have offsets in some arbitrary order
+        // but the buffer contains the actual strings in the order they were inserted
+        // so we need to build offsets for the strings in the buffer in order
+        // then append short strings, if any, and then build the StringArray
+        // TODO a picture would be nice here
+        let Self {
+            map,
+            map_size: _,
+            mut buffer,
+            random_state: _,
+            hashes_buffer: _,
+        } = self;
+
+        // Sort all headers so that long strings come first, in offset order
+        // followed by short strings ordered by value
+        let mut headers = map.into_iter().collect::<Vec<_>>();
```

Comment:
this sort shows up on my profiles. I think we could potentially improve it by remembering the sort order of the strings so we don't have to derive it from the contents of the hash tabklw

Reply:
fixed in [a101b62](https://github.com/apache/arrow-datafusion/pull/8849/commits/a101b62557d0c0aaa0402256cc4ca249995d5b04)